Source/WebCore/platform/graphics/WidthIterator.cpp

```diff
@@ -268,40 +297,86 @@ inline void WidthIterator::advanceInternal(TextIterator& textIterator, GlyphBuff
             continue;
         }
 #endif
-        const GlyphData& glyphData = m_font.glyphDataForCharacter(character, rtl);
-        Glyph glyph = glyphData.glyph;
+        const GlyphData& glyphData = m_font.glyphDataForCharacter(character, false, FontVariant::NormalVariant);
+        font = glyphData.font ? glyphData.font : &primaryFont;
+        auto shouldSynthesizeCharacter = shouldSynthesize(dontSynthesizeSmallCaps, font, character, capitalizedCharacter, fontVariantCaps, engageAllSmallCapsProcessing);
+
+        if (synthesizedFont || shouldSynthesizeCharacter) {
+            if (capitalizedCharacter) {
+                nextIsSmallCaps = true;
+                characterToWrite = capitalizedCharacter.value();
+            } else
+                nextIsSmallCaps = engageAllSmallCapsProcessing;
+        }
+        if (rtl)
+            characterToWrite = u_charMirror(characterToWrite);
+
+        auto rangeFont = fontForRange(lastFontData, synthesizedFont, smallSynthesizedFont, isSmallCaps);
+        auto nextRangeFont = rangeFont;
+
+        // Checking if we should start a new range
+        if (font != lastFontData || nextIsSmallCaps != isSmallCaps) {
+            commitCurrentFontRange(glyphBuffer, rangeBuffer, lastGlyphCount, currentCharacterIndex, rangeFont, primaryFont, widthOfCurrentFontRange, charactersTreatedAsSpace);
+
+            if (font != lastFontData) {
+                synthesizedFont = nullptr;
+                smallSynthesizedFont = nullptr;
+                nextIsSmallCaps = false;
+            }
+            if (shouldSynthesizeCharacter) {
+                synthesizedFont = &font->noSynthesizableFeaturesFont();
+                smallSynthesizedFont = synthesizedFont->smallCapsFont(m_font.fontDescription());
+                nextIsSmallCaps = true;
+            }
+            nextRangeFont = fontForRange(font, synthesizedFont, smallSynthesizedFont, nextIsSmallCaps);
+            rangeBuffer.clear();
+            indexOfFontTransition = currentCharacterIndex;
+        }
+
+        if (!synthesizedFont && shouldSynthesizeCharacter) {
+            // Rather than synthesize each character individually, we should synthesize the entire "run" if any character requires synthesis.
+            synthesizedFont = &font->noSynthesizableFeaturesFont();
+            smallSynthesizedFont = synthesizedFont->smallCapsFont(m_font.fontDescription());
+            nextIsSmallCaps = true;
+            textIterator.reset(indexOfFontTransition);
+            continue;
+        }
+
+        Glyph glyph;
+        if (synthesizedFont || shouldSynthesizeCharacter)
+            glyph = nextRangeFont->glyphForCharacter(characterToWrite);
+        else
+            glyph = font->glyphForCharacter(characterToWrite);
```

Comment:
this is a lot of code to embed in the middle of an already-big function that isn't going to be used most of the time. Is it possible to refactor this to move most of this logic to separate functions? Just to try to keep `advanceInternal()` somewhat readable?